applications/ParticleMechanicsApplication/custom_constitutive/johnson_cook_thermal_plastic_3D_law.cpp

```diff
@@ -0,0 +1,476 @@
+
+			// Store plastic deformation quantities
+			mEnergyDissipated += gamma / GetSqrt6() / MaterialProperties[DENSITY] * (mYieldStressOld + yield_stress);
+			mEquivalentPlasticStrainOld = predicted_eps;
+			mPlasticStrainRateOld = predicted_eps_rate;
+			mTemperatureOld = predicted_temperature;
+			mYieldStressOld = yield_stress;
+			mHardeningRatio = yield_stress / mYieldStressVirgin;
+		}
+		else
+		{
+			if (yield_stress / mYieldStressVirgin > yield_stress_failure_ratio)
+			{
+				stress_deviatoric_converged = stress_deviatoric_trial;
+			}
+			else
+			{
+				// Particle has failed. It can only take compressive volumetric stresses!
+				stress_deviatoric_converged.clear();
+				if (stress_hydrostatic_new > 0.0) stress_hydrostatic_new = 0.0;
+				mHardeningRatio = 0.0;
+			}
+		}
+		// Update equivalent stress
+		Matrix stress_converged = stress_deviatoric_converged + stress_hydrostatic_new * identity;
+		mEquivalentStress = std::sqrt(3.0 / 2.0 * CalculateMatrixDoubleContraction(stress_deviatoric_converged));
+
+		// Store stresses and strains
+		MakeStrainStressVectorFromMatrix(stress_converged, StressVector);
+		mStrainOld = StrainVector;
+
+		// Udpdate internal energy
+		for (size_t i = 0; i < stress_converged.size1(); ++i) {
+			for (size_t j = 0; j < stress_converged.size2(); ++j) {
+				mEnergyInternal += 0.5 * (stress_converged(i, j) + stress_old(i, j)) * strain_increment(i, j) / MaterialProperties[DENSITY];
+			}
+		}
+	}
+
+
+	int JohnsonCookThermalPlastic3DLaw::Check(const Properties& rMaterialProperties, const GeometryType& rElementGeometry, const ProcessInfo& rCurrentProcessInfo)
+	{
+		const int check_base = BaseType::Check(rMaterialProperties, rElementGeometry, rCurrentProcessInfo);
+
+		KRATOS_ERROR_IF (JC_PARAMETER_A.Key()==0 || rMaterialProperties[JC_PARAMETER_A] < 0.0) << "JC_PARAMETER_A has key zero or invalid value (expected positive number ~500MPa)" << std::endl;
+		KRATOS_ERROR_IF (JC_PARAMETER_B.Key()==0 || rMaterialProperties[JC_PARAMETER_B] < 0.0) << "JC_PARAMETER_B has key zero or invalid value (expected positive number ~500MPa)" << std::endl;
+		KRATOS_ERROR_IF (JC_PARAMETER_C.Key()==0 || rMaterialProperties[JC_PARAMETER_C] < 0.0) << "JC_PARAMETER_C has key zero or invalid value (expected positive number ~0.01)" << std::endl;
+		KRATOS_ERROR_IF (JC_PARAMETER_n.Key()==0 || rMaterialProperties[JC_PARAMETER_n] < 0.0) << "JC_PARAMETER_n has key zero or invalid value (expected positive number ~0.25)" << std::endl;
+		KRATOS_ERROR_IF (REFERENCE_STRAIN_RATE.Key()==0 || rMaterialProperties[REFERENCE_STRAIN_RATE] <= 0.0) << "REFERENCE_STRAIN_RATE has key zero or invalid value (expected positive number ~1.0)" << std::endl;
+		KRATOS_ERROR_IF (TAYLOR_QUINNEY_COEFFICIENT.Key()==0 || rMaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] < 0.0) << "TAYLOR_QUINNEY_COEFFICIENT has key zero or invalid value (expected positive number ~0.9)" << std::endl;
```

Comment:
You have this already, so the above ifs can be removed?

Reply:
I'm not sure what you mean? Could you explain a bit more please? This checks that the material parameter inputs are valid (thanks to your law implementations btw 👍 ).

Reply:
I mean the following: checking if for all material points is pretty expensive, so we want to reduce that as much as possible. I was thinking that in the beginning, if we check the parameters of TAYLOR_QUINNEY_COEFFICIENT to be less than 0.0, why don't we check if the value is zero or positive there too. Then, assign a `double multiplier_`, which is zero if your thermal is not used and 1.0 if your thermal softening is used. Every time you want to compute the hardening law, you can just multiply the hardening by the `multiplier_` without even checking it. I am not sure if I am explaining it nicely.

Reply:
Ah I get you - it turns out that multiplying by the zero value of TAYLOR_QUINNEY_COEFFICIENT 
 achieves the desired result (without having to use the `multiplier_`  approach).